api_tutorials/bqsr_cnn.py

```diff
@@ -0,0 +1,997 @@
+
+	Returns:
+		The Keras model
+	'''
+	with open(semantics_json, 'r') as infile:
+		semantics = json.load(infile)
+
+	if 'input_tensor_name' in semantics:
+		args.tensor_name = semantics['input_tensor_name']
+		args.window_size = semantics['window_size']
+		args.read_limit = semantics['read_limit']
+		assert(len(args.input_symbols) == len(semantics['input_tensor_name']))
+		for key in args.input_symbols:
+			assert(args.input_symbols[key] == semantics['input_tensor_name'][key])
+
+	if 'input_annotations' in semantics:
+		args.annotations = semantics['input_annotations']
+		args.annotation_set = semantics['input_annotation_set']
+
+	if 'channels_last' in semantics:
+		args.channels_last = semantics['channels_last']
+		if args.channels_last:
+			K.set_image_data_format('channels_last')
+		else:
+			K.set_image_data_format('channels_first')
+				
+	args.input_symbols = semantics['input_symbols']
+	args.labels = semantics['output_labels']
+
+	weight_path_hd5 = os.path.join(os.path.dirname(semantics_json),semantics['architecture'])
+	model = load_model(weight_path_hd5, custom_objects=bqsr_get_metric_dict(args.labels))
+	model.summary()
+	return model
+
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# ~~~~~~~ Writing and Generating Tensors ~~~~~~~~~~~~~~~~
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+def write_base_recalibrate_tensors(args, include_annotations=True):
+	"""Create tensors structured as tensor map of read quality.
+
+	Defines true bases as those not in args.db_snp, where read and reference agree.
+	False bases are sites thos not in args.db_snp where the read and the reference do NOT agree. 
+
+	Arguments
+		args.data_dir: directory where tensors will live. Created here and filled with
+			subdirectories of test, valid and train, each containing
+			subdirectories for each label with tensors stored as hd5 files.
+		args.bam_file: BAM or BAMout file where the aligned reads are stored
+		args.ignore_vcf: VCF file with sites of known variation, from NIST, DBSNP etc.
+		args.bed_file: Intervals of interest
+		args.window_size: Size of sequence window around variant (width of the tensor)
+		args.chrom: Only write tensors from this chromosome (optional, used for parallelization)
+		args.start_pos: Only write tensors after this position (optional, used for parallelization)
+		args.end_pos: Only write tensors before this position (optional, used for parallelization)
+	"""	
+	# NEED to label, set max size, go to only p-hot 4xwindow size, 
+	# eliminate any known variant covering read
+	# iterate over vcf
+	# check between variants for reads
+	print('Writing base recalibration tensors from tensor channel map.')
+	stats = Counter()
+
+	samfile = pysam.AlignmentFile(args.bam_file, "rb")	
+	vcf_ram = vcf.Reader(open(args.ignore_vcf, 'r'))
+	
+	bed_dict = bqsr_bed_file_to_dict(args.bed_file)
+	record_dict = SeqIO.to_dict(SeqIO.parse(args.reference_fasta, "fasta"))
+
+	margin = args.window_size/2
```

Comment:
It looks like pysam's fetch() captures all reads that intersect the [start,end] interval. So the current code is capturing many reads that have NIST variants. We should get somewhat better empirical quality if we get rid of the `/2`.